### PR TITLE
RISCV: Support pm cpu ops for SMP

### PR DESCRIPTION
**Pull request by quic-lingutla**
Add pm cpu ops to call the platform specific implementations for bringing up secondary cores.

### CODE REVIEW

File: arch/riscv/core/smp.c [hunk 1]
```diff
@@ -27,6 +32,15 @@ void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
 	riscv_cpu_sp = Z_KERNEL_STACK_BUFFER(stack) + sz;
 	riscv_cpu_wake_flag = _kernel.cpus[cpu_num].arch.hartid;
 
+#ifdef CONFIG_PM_CPU_OPS
+	uintptr_t entry = (uintptr_t)&__start;
+
+	if (pm_cpu_on(cpu_num, entry)) {
```
Comment (carlocaione) [inline arch/riscv/core/smp.c:38]: I don't get it. Is RISCV supporting PSCI and SMCCC now? What PM CPU ops driver is this using?

**Review by carlocaione — CHANGES_REQUESTED** [review]

File: arch/riscv/core/smp.c (hunk 1 above)
Comment (quic-lingutla) [reply arch/riscv/core/smp.c:38]: I am not sure why PM CPU ops tightly coupled with PSCI or SMCC.  RISCV SMP's arch_start_cpu() assumes secondary core is already powered up and looping at boot_secondary_core, but some platforms needs to do initialization (either via opensbi or regular MMIO to enable clocks, etc,.) to bring up secondary cores.

As of now, RISCV doesn't have any PM CPU Ops driver, but platforms can implement weak functions to bringup secondary cores.

Comment (carlocaione) [reply arch/riscv/core/smp.c:38]: My point is exactly that RISCV doesn't have any PM CPU ops driver in tree so you are adding this for some out-of-tree platform?

Comment (quic-lingutla) [reply arch/riscv/core/smp.c:38]: Right, as of now the platform is out of tree.

Comment (carlocaione) [reply arch/riscv/core/smp.c:38]: ok, do you plan to upstream it? I'm not willing to push code into the tree that is basically dead code without any user for it.

Comment (quic-egmc) [reply arch/riscv/core/smp.c:38]: Summarizing the offline conversation we had...

This code is in support of a microcontroller core embedded in a larger SoC design so there's not a clear path to share a buildable platform.  In this particular platform the driver implementing the CPU Ops talks to some proprietary power/clock/reset controlling hardware.

As a separate discussion supporting Zephyr-on-SBI on an upstream platform (at least QEMU) would be really interesting.

File: arch/riscv/core/smp.c [hunk 2]
```diff
@@ -18,6 +19,10 @@ volatile struct {
 volatile uintptr_t riscv_cpu_wake_flag;
 volatile void *riscv_cpu_sp;
 
+#ifdef CONFIG_PM_CPU_OPS
+extern void __start(void);
```
Comment (carlocaione) [inline arch/riscv/core/smp.c:23]: should this be `CONFIG_KERNEL_ENTRY`?

Comment (quic-lingutla) [reply arch/riscv/core/smp.c:23]: Are you referring to the above #ifdef check ? or function declaration ?

If its for function declaration, 
I am not sure, if we have any helpers to declare function like: 
extern void CONFIG_KERNEL_ENTRY(void);

Comment (carlocaione) [reply arch/riscv/core/smp.c:23]: uhm, I see. This should really be fixed somehow, it is really useless to have `CONFIG_KERNEL_ENTRY` and not being able to use it in C code (but only in the linker script).

File: arch/riscv/core/smp.c (hunk 1 above)
Comment (carlocaione) [inline arch/riscv/core/smp.c:38]:
```suggestion
	if (pm_cpu_on(cpu_num, (uintptr_t)&__start)) {
```

File: arch/riscv/core/smp.c [hunk 3]
```diff
@@ -18,6 +19,10 @@ volatile struct {
 volatile uintptr_t riscv_cpu_wake_flag;
 volatile void *riscv_cpu_sp;
 
+#ifdef CONFIG_PM_CPU_OPS
```
Comment (carlocaione) [inline arch/riscv/core/smp.c:22]: This `ifdef` is not needed.

**Review by carlocaione — CHANGES_REQUESTED** [review]

File: arch/riscv/core/smp.c (hunk 3 above)
Comment (quic-lingutla) [reply arch/riscv/core/smp.c]: Done. Thanks

**Review by carlocaione — APPROVED** [review]

**Review by fkokosinski — APPROVED** [review]

**Pull request merged**